{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a8d9067-9efa-48aa-b878-509dc345fa29",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 01 Trajectory Analysis by Palantir   refer to https://github.com/dpeerlab/Palantir/blob/master/notebooks/Palantir_sample_notebook.ipynb\n",
    "import scanpy as sc\n",
    "import os\n",
    "import pandas as pd\n",
    "from cellrank.tl.kernels import CytoTRACEKernel,ConnectivityKernel\n",
    "from cellrank.tl.estimators import GPCCA\n",
    "import scanpy.external as sce\n",
    "import palantir\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import warnings\n",
    "from numba.core.errors import NumbaDeprecationWarning\n",
    "\n",
    "warnings.filterwarnings(action=\"ignore\", category=NumbaDeprecationWarning)\n",
    "warnings.filterwarnings(\n",
    "    action=\"ignore\", module=\"scanpy\", message=\"No data for colormapping\"\n",
    ")\n",
    "\n",
    "adata=sc.read(\"/path/to/trophoblast.h5ad\")\n",
    "\n",
    "ck = ConnectivityKernel(adata)\n",
    "ck.compute_transition_matrix()\n",
    "g_pv = GPCCA(ck)\n",
    "g_pv.compute_schur(n_components=20)\n",
    "g_pv.plot_spectrum(real_only=True)\n",
    "g_pv.compute_macrostates(n_states=10, cluster_key=\"subtype\")\n",
    "g_pv.plot_macrostates(discrete=True, legend_loc=\"right\", size=100, basis=\"X_umap\")\n",
    "g_pv.set_terminal_states_from_macrostates(['eEVT','SCT_b','iEVT',\"SCT_a\"])\n",
    "g_pv._set_initial_states_from_macrostates(\"VCT_div\")\n",
    "g_pv.compute_absorption_probabilities()\n",
    "g_pv.plot_absorption_probabilities(same_plot=False)\n",
    "\n",
    "start_cell=trophoblast.obs.initial_states_probs.argmax()\n",
    "start_cell=trophoblast.obs.index[root_cell]\n",
    "\n",
    "end_cell=adata.obs.terminal_states_probabilities.idxmax(axis=0)\n",
    "terminal_states=end_cell.idxmax(axis=0)\n",
    "\n",
    "dm_res = palantir.utils.run_diffusion_maps(adata, n_components=20,pca_key='X_pca_harmony')\n",
    "ms_data = palantir.utils.determine_multiscale_space(adata)\n",
    "imputed_X = palantir.utils.run_magic_imputation(adata)\n",
    "\n",
    "pr_res = palantir.core.run_palantir(\n",
    "    adata, start_cell, num_waypoints=500, terminal_states=terminal_states)\n",
    "\n",
    "\n",
    "pr_res.branch_probs.columns = adata.obs.loc[pr_res.branch_probs.columns,'subtype']\n",
    "adata.obs['palantir_pseudotime']=pr_res.pseudotime\n",
    "sc.pl.umap(adata,color='palantir_pseudotime',cmap='viridis')\n",
    "\n",
    "## plot trajectories on UMAP\n",
    "palantir.plot.plot_trajectories(adata, pseudotime_interval=(0, .9))\n",
    "palantir.plot.plot_trajectories(adata, cell_color = \"branch_selection\", pseudotime_interval=(0, .9))\n",
    "\n",
    "masks = palantir.presults.select_branch_cells(adata, q=.01, eps=.01)\n",
    "palantir.plot.plot_branch_selection(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c64775b-0ede-4568-80ae-695def9f09fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 02 Composition analysis of STOMICS\n",
    "\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "\n",
    "df_bv=pd.read_table(\"/path/to/distance_to_bv.tsv\",sep='\\t',index_col=0,header=0)\n",
    "df_interface=pd.read_table(\"/path/to/distance_to_interface.tsv\",sep='\\t',index_col=0,header=0)\n",
    "bins=[-4000,-2000,-1500,-1000,-500,-250,0,250,500,1000,1500,2000,4000]\n",
    "## interface\n",
    "df_interface['distance_bin'] = pd.cut(df_interface['signed_distance_interface'], bins)\n",
    "composition = df_interface.groupby(['distance_bin', 'celltype']).size().reset_index(name='count')\n",
    "composition['proportion'] = composition.groupby('distance_bin')['count'].transform(lambda x: x / x.sum())\n",
    "plt.figure(figsize=(10, 5))\n",
    "composition_pivot = composition.pivot(index=\"distance_bin\", columns=\"celltype\", values=\"proportion\").fillna(0)\n",
    "composition_pivot.plot(kind=\"bar\", stacked=True, colormap=\"tab20\", figsize=(10, 5))\n",
    "plt.xlabel(\"Distance from interface\")\n",
    "plt.ylabel(\"Proportion of Cell Type\")\n",
    "plt.xticks(rotation=45)\n",
    "\n",
    "#BV\n",
    "bins=[-1000,-100,0,50,100,150,200,300,500,1000]\n",
    "df_bv['distance_bin'] = pd.cut(df_bv['signed_distance_bv'], bins)\n",
    "composition = df_bv.groupby(['distance_bin', 'celltype']).size().reset_index(name='count')\n",
    "composition['proportion'] = composition.groupby('distance_bin')['count'].transform(lambda x: x / x.sum())\n",
    "plt.figure(figsize=(10, 5))\n",
    "composition_pivot = composition.pivot(index=\"distance_bin\", columns=\"celltype\", values=\"proportion\").fillna(0)\n",
    "composition_pivot.plot(kind=\"bar\", stacked=True, colormap=\"tab20\", figsize=(10, 5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f56620f1-95d0-4a77-b6e4-c84cbf5800ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 03 Differential Expressed Genes in EVT subtypes\n",
    "\n",
    "\n",
    "\n",
    "def plot_scatter(data,x,y,group,degs,adjust=True,**kwargs):\n",
    "    ax=plt.figure(figsize=(10,10))\n",
    "    ax=sns.scatterplot(data=data,x=x,y=y,hue=group,size =2, linewidth=0, **kwargs)\n",
    "    labeled_x,labeled_y=data.loc[degs,x],data.loc[degs,y]\n",
    "    texts = [plt.text(labeled_x[i], labeled_y[i], degs[i], fontsize='medium') for i in range(len(labeled_x))]\n",
    "    if adjust:\n",
    "        adjust_text(texts, labeled_x.values, labeled_y.values, arrowprops=dict(arrowstyle='-', color='black'),force_text=(2,2),expand_text=(0.5,0.5))\n",
    "    plt.xlabel('scaled expression')                   # Set x-axis label\n",
    "    plt.ylabel('Log2FC')\n",
    "    return ax\n",
    "\n",
    "\n",
    "evt=adata[adata.obs.subtype.isin(['pEVT','eEVT','iEVT']),:]\n",
    "\n",
    "## Compute DEGs\n",
    "sc.tl.rank_genes_groups(adata,groupby='subtype',method='wilcoxon',refenece='iEVT')\n",
    "deg1=sc.get.rank_genes_groups_df(adata,group=['pEVT'])\n",
    "deg2=sc.get.rank_genes_groups_df(adata,group=['eEVT'])\n",
    "\n",
    "df=pd.DataFrame(index=evt.var_names)\n",
    "df[['padj_pEVT','scores_PvsI']]=deg1.loc[df.index,['pvals_adj','scores']]\n",
    "df[['padj_eEVT','scores_EvsI']]=deg2.loc[df.index,['pvals_adj','scores']]\n",
    "\n",
    "genestolabel=['PGF','SEMA4B','ELMO1','PLXDC2','FN1',\n",
    "      \"SLIT2\",\"BMP5\",'TCF4',\"GATA4\",\"ZEB2\",\n",
    "     \"DOCK5\",\"DOCK1\",\"MAP4K4\",\"GRK5\",\"ACTN1\",\"ITGA1\",\"VAV2\",\n",
    "     \"ITGB1\",\"ANXA1\",\"IFI6\",\"HLA-G\",\"TIMP2\",\"TIMP3\",'GRB2','AOC1','B2M','PRG2','SERPINE1','SERPINE2','GJA1',\n",
    "     \"CD63\",'CD59','FTL','KRT8','IGF2','SEMA5A','HTRA1','HES2']\n",
    "df['group']='genome'\n",
    "df.loc[df['score_pEVT'].gt(20),'group']='pEVT'\n",
    "df.loc[df['score_eEVT'].gt(20),'group']='eEVT'\n",
    "df.loc[(df['score_eEVT'].lt(-20))&(df['score_pEVT'].lt(-20)),'group']='iEVT'\n",
    "\n",
    "palette=['springgreen','dodgerblue','scarlet','lightgrey']\n",
    "plot_scatter(data=df,x=\"scores_PvsI\",y='scores_EvsI',adjust=True,degs=genestolabel,group=\"group\",palette=sns.color_palette(palette))\n",
    "plt.ylim((-60,60))\n",
    "plt.xlim((-20,40))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ce911c7-6c2e-4ba2-af3c-6d695fff58cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 04 Scoring Pseudtovascularization of EVTs\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pygam import s, LinearGAM\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "from scipy.stats import zscore\n",
    "\n",
    "plt.rcParams['pdf.fonttype']=42\n",
    "\n",
    "evt=adata[adata.obs.subclass=='EVT',:] ## subset EVTs from STOMICS\n",
    "distances=pd.read_table(\"/path/to/distance_bv.tsv\",sep=\"\\t\",index_col=0,header=0)  ## distance computed on STOMICS\n",
    "distances=distances.loc[[i for i in evt.obs_names if i in distances.index]]\n",
    "degs=pd.read_csv(\"/path/to/marker_mVEC.tsv\",sep=\"\\t\",index_col=0) ## load mVEC markers from snRNA-seq \n",
    "degs = degs.sort_values(by=\"scores\", ascending=False).head(100).index\n",
    "degs=[i for i in degs if 'RPS' not in i and 'RPL' not in i and \"MT-\" not in i and i in evt.var_names] ## remove ribosomal and mitochondria genes\n",
    "\n",
    "#genes=['PECAM1','MMP1','CDH5','CLEC14A','VWF','CD34','MCAM','THSD1']\n",
    "sc.tl.score_genes(evt,genes,score_name='aEC_score')\n",
    "distances['aEC_scaled_score']=zscore(evt.obs['aEC_score'],ddof=1)\n",
    "bins = list(range(0, 301, 25))\n",
    "distances['distance_bin'] = pd.cut(distances['signed_distance'], bins)\n",
    "distances=distances.dropna()\n",
    "\n",
    "sns.boxplot(distances,x='distance_bin',y='aEC_scaled_score',fliersize=0)\n",
    "\n",
    "\n",
    "X,y=distances['signed_distance'],distances['aEC_scaled_score']\n",
    "gam = LinearGAM(s(0, n_splines=10, lam=5)).fit(X, y)\n",
    "predictions = gam.predict(np.linspace(0,300,50))\n",
    "plt.plot(mp.linspace(0,300,50), predictions, color='red')\n",
    "confint = gam.confidence_intervals(XX)\n",
    "plt.fill_between(XX, confint[:, 0], confint[:, 1], alpha=0.3, label='95% CI')\n",
    "plt.ylim(-2,1.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1229db00-1800-44ed-9c11-3762ddb01df9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 05 Invasive score modeling\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn import train_test_split\n",
    "from sklearn.linear_model import LassoCV, Lasso\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import pickle\n",
    "import scipy.stats as ss\n",
    "## data preprocessing\n",
    "df_interface=pd.read_table(\"/path/to/distance_interface.tsv\",sep=\"\\t\",index_col=0,header=0)\n",
    "\n",
    "expr_df = pd.DataFrame(evt.X.toarray(),index=evt.obs_names,columns=evt.var_names)\n",
    "genes=evt.var_names\n",
    "cell_id=evt.obs.loc[evt.obs.celltype.isin(['EVT','iEVT','EVTpro']),:].index\n",
    "masked=df_interface.loc[(df_interface['signed_distance_interface']>0),:].index\n",
    "cell_id=np.intersect1d(cell_id,masked)\n",
    "\n",
    "train_cells, test_cells = train_test_split(cell_id,test_size=0.5,random_state=42)\n",
    "train_df,test_df=expr_df.loc[train_cells],expr_df.loc[test_cells]\n",
    "train_df['distance_to_interface']=df_interface.loc[train_cells,'distance_to_interface']\n",
    "train_df=train_df.sort_values(by='distance_to_interface', ascending=True).reset_index(drop=True)\n",
    "N=10\n",
    "train_df['group']=train_df.index//N\n",
    "train_mean = train_df.groupby('group')[genes+['distance_to_interface']].mean(numeric_only=True)\n",
    "## restrict to genes highly expressed in EVTs\n",
    "mks=pd.read_table(\"/path/to/EVT_marker.tsv\",sep=\"\\t\",header=0)\n",
    "candidates=mks.loc[(mks.scores.gt(2)),'names'].values\n",
    "candidates=[i for i in candidates if not (i.startswith('RPL') | i.startswith('MT-') | i.startswith('RPS')) and i in genes]\n",
    "\n",
    "## Feature selection by LASSO\n",
    "X = StandardScaler().fit_transform(train_mean[candidates])\n",
    "y=train_mean['distance_to_interface'].values\n",
    "lasso = Lasso(alpha=50).fit(X,y)\n",
    "coef_lasso = pd.Series(lasso.coef_, index=candidates).sort_values(ascending=True)\n",
    "print((coef_lasso!=0).sum())\n",
    "with open(\"model.pkl\", \"rb\") as f:\n",
    "    pickle.dump(lasso,f)\n",
    "\n",
    "## Model Validation \n",
    "X_test=StandardScaler().fit_transform(test_df[candidates])\n",
    "y_pred=lasso.predict(X_test)\n",
    "print(ss.spearmanr(df_interface.loc[test_cells,'signed_distance_interface'].values,y_pred)\n",
    "\n",
    "## Invasive Scoring by snRNA-seq \n",
    "\n",
    "adata=sc.read(\"/path/to/trophoblast.h5ad\")\n",
    "snRNA_ad=adata[adata.obs.major_class=='EVT',:]\n",
    "snRNA_input = pd.DataFrame(snRNA_ad.X.toarray(),index=snRNA_ad.obs_names,columns=snRNA_ad.var_names)\n",
    "snRNA_input = StandardScaler().fit_transform(snRNA_input[candidates])\n",
    "pred=pd.DataFrame(lasso.predict(snRNA_input),index=snRNA_ad.obs_names,columns='raw_pred')\n",
    "pred['iScore']=(pred['raw_pred'] - pred['raw_pred'].mean()) / pred['raw_pred'].std()\n",
    "## visualization\n",
    "snRNA_ad.obs['iScore']=pred['iScore']\n",
    "sns.boxplot(data=snRNA_ad.obs, x='subtype',y='iScore', hue='subtype',fliersize=0)\n",
    "sc.pl.umap(snRNA_ad,colour=['iScore'],cmap='viridis')\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
