{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8e7bbef-5501-44b4-b550-5f65c8dd2005",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 01 Trajectory Inference\n",
    "import scanpy as sc\n",
    "import os\n",
    "import pandas as pd\n",
    "from cellrank.tl.kernels import CytoTRACEKernel,ConnectivityKernel\n",
    "from cellrank.tl.estimators import GPCCA\n",
    "import scanpy.external as sce\n",
    "import palantir\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import warnings\n",
    "from numba.core.errors import NumbaDeprecationWarning\n",
    "\n",
    "warnings.filterwarnings(action=\"ignore\", category=NumbaDeprecationWarning)\n",
    "warnings.filterwarnings(\n",
    "    action=\"ignore\", module=\"scanpy\", message=\"No data for colormapping\"\n",
    ")\n",
    "\n",
    "adata=sc.read(\"/path/to/dsc.h5ad\")\n",
    "\n",
    "ck = ConnectivityKernel(adata)\n",
    "ck.compute_transition_matrix()\n",
    "g_pv = GPCCA(ck)\n",
    "g_pv.compute_schur(n_components=20)\n",
    "g_pv.plot_spectrum(real_only=True)\n",
    "g_pv.compute_macrostates(n_states=10, cluster_key=\"subtype\")\n",
    "g_pv.plot_macrostates(discrete=True, legend_loc=\"right\", size=100, basis=\"X_umap\")\n",
    "g_pv.set_terminal_states_from_macrostates([\"DSC4\",\"DSC3\"])\n",
    "g_pv._set_initial_states_from_macrostates(\"DSC0\")\n",
    "g_pv.compute_absorption_probabilities()\n",
    "g_pv.plot_absorption_probabilities(same_plot=False)\n",
    "\n",
    "start_cell=trophoblast.obs.initial_states_probs.argmax()\n",
    "start_cell=trophoblast.obs.index[root_cell]\n",
    "\n",
    "end_cell=adata.obs.terminal_states_probabilities.idxmax(axis=0)\n",
    "terminal_states=end_cell.idxmax(axis=0)\n",
    "\n",
    "dm_res = palantir.utils.run_diffusion_maps(adata, n_components=5,pca_key='X_pca_harmony')\n",
    "ms_data = palantir.utils.determine_multiscale_space(adata)\n",
    "imputed_X = palantir.utils.run_magic_imputation(adata)\n",
    "\n",
    "pr_res = palantir.core.run_palantir(\n",
    "    adata, start_cell, num_waypoints=500, terminal_states=terminal_states)\n",
    "\n",
    "pr_res.branch_probs.columns = adata.obs.loc[pr_res.branch_probs.columns,'subtype']\n",
    "adata.obs['palantir_pseudotime']=pr_res.pseudotime\n",
    "sc.pl.umap(adata,color='palantir_pseudotime',cmap='viridis')\n",
    "\n",
    "\n",
    "## plot trajectories on UMAP\n",
    "palantir.plot.plot_trajectories(adata, pseudotime_interval=(0, .9))\n",
    "palantir.plot.plot_trajectories(adata, cell_color = \"branch_selection\", pseudotime_interval=(0, .9))\n",
    "\n",
    "\n",
    "##stream plot\n",
    "import math\n",
    "from streamutils import *\n",
    "time_key=\"palantir_pseudotime\"\n",
    "a=Gradient_calculator(adata,obsm_key='X_umap',pseudotime_key=time_key)\n",
    "plt.rcParams['savefig.bbox'] = 'tight'\n",
    "a.calculate_p_mass(n_grid=20)\n",
    "a.suggest_mass_thresholds()\n",
    "a.calculate_mass_filter(min_mass=10,plot=True)\n",
    "plt.style.use(\"seaborn\")\n",
    "a.transfer_data_into_grid()\n",
    "a.calculate_gradient()\n",
    "a.plot_reference_flow_on_grid(density=5)\n",
    "a.plot_dev_flow_on_grid()\n",
    "a.visualize_results(plot_type=\"stream\",s=11,cmap=\"rainbow\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "134dcdba-04ed-43dd-9909-f0722d099475",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 02 Composition changes over time\n",
    "\n",
    "sc.pp.neighbors(adata,n_neighbors=40, n_pcs=50, use_rep='X_pca')\n",
    "adata.obs['Days']=adata.obs['GW']*7.0\n",
    "wk = WOTKernel(adata, time_key=\"Days\")\n",
    "wk.compute_initial_growth_rates(organism=\"human\", key_added=\"growth_rate_init\")\n",
    "sc.pl.scatter(adata, color=\"growth_rate_init\", legend_loc=\"right\", basis=\"X_umap\", s=10)\n",
    "wk.compute_transition_matrix(growth_iters=3, growth_rate_key=\"growth_rate_init\", last_time_point=\"connectivities\")\n",
    "groups=['DSC0','DSC1','DSC2','DSC3','DSC4']\n",
    "ax = wk.plot_single_flow(cluster_key=\"subtype\",cluster=\"DSC0\",clusters=groups,time_key=\"Days\",min_flow=0,\n",
    "                         xticks_step_size=4,show=False,dpi=100)\n",
    "locs, labels = plt.xticks()\n",
    "ax.set_xticks(locs)\n",
    "ax.set_xticklabels(labels, rotation=90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad76c1f1-5bb7-4ef7-b1c6-50044bd9b577",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 03 Differential Expression Analysis \n",
    "\n",
    "# Path A vs Path B\n",
    "import seaborn as sns\n",
    "\n",
    "adata.obs['Path']=adata.obs['subtype'].astype('str').map({'DSC0':'None', 'DSC1':\"A\", 'DSC3':\"A\", 'DSC2':\"B\", 'DSC3':\"B\"})\n",
    "\n",
    "\n",
    "sc.tl.rank_genes_groups(adata,groupby=\"Path\",groups=[\"A\",\"B\"],\n",
    "                        method=\"wilcoxon\",reference=\"B\",n_genes=30)\n",
    "\n",
    "df=sc.get.rank_genes_groups_df(adata,group=['A'])\n",
    "mean_exp=pd.DataFrame(np.array(adata.X.mean(axis=0)).ravel(),index=adata.var_names,columns=['basemean'])\n",
    "df['basemean']=mean_exp.loc[df.index,'basemean']\n",
    "df['group']='BG'\n",
    "df.loc[df['scores'].gt(5),'group']=\"A\"\n",
    "df.loc[df['scores'].lt(-5),'group']=\"B\"\n",
    "\n",
    "genestolabel=['IGFBP1','LGR4','TIMP3','SEMA3A','DOCK1','PRLR',\"PRL\",'DOCK4','SAT1',\"ITGA1\",\"ITGB1\",\"THSD1\",'LUM',\n",
    "     \"VIM\",\"B2M\",'CD63','TIMP1','TIMP2','CD81','IGFBP2','HLA-C','HLA-A','HLA-B','CD248','TPM2','PPIA','S100A11',\n",
    "     'CD28','TPT1','SERPINF1','PPIB','C1R','OAZ1','CD59','IGFBP5','GJA1','IGFBP4']\n",
    "ax = sns.scatterplot(x=\"basemean\", # Horizontal axis\n",
    "            y='logfoldchanges', # Vertical axis\n",
    "           data=df, # Data source # Don't fix a regression line\n",
    "           size = 1,hue='group',alpha=0.7,palette='coolwarm_r',linewidth=0,hue_order=[\"B\",\"BG\",\"A\"]), # size and dimension\n",
    "\n",
    "plt.xlabel('Mean exp')\n",
    "plt.ylabel('Log2FC(Path A vs B)')\n",
    "\n",
    "def label_point(x, y, val, ax):\n",
    "    a = pd.concat({'x': x, 'y': y, 'val': val}, axis=1)\n",
    "    for i, point in a.iterrows():\n",
    "        ax.text(point['x']+.02, point['y'], str(point['val']),size=8)\n",
    "label_point(df.loc[genestolabel,'basemean'], df.loc[genestolabel,'logfoldchanges'], df.loc[genestolabel,'Gene name'], plt.gca())\n",
    "plt.ylim((-5,5))\n",
    "\n",
    "## Between DSC3 subclusters\n",
    "\n",
    "dsc3=adata[adata.obs.subtype==\"DSC3\",:]\n",
    "sc.tl.louvain(dsc3,resolution=0.1)\n",
    "dec31=dsc3.obs.loc[dsc3.obs.louvain==\"0\",:].index\n",
    "dec32=dsc3.obs.loc[dsc3.obs.louvain==\"1\",:].index\n",
    "adata.obs['subclusters']=adata.obs['subtype'].astype('str')\n",
    "adata.obs.loc[dec31,\"subclusters\"]=\"DSC3_1\"\n",
    "adata.obs.loc[dec32,\"subclusters\"]=\"DSC3_2\"\n",
    "sc.tl.rank_genes_groups(dec,groupby=\"subclusters\",groups=[\"DSC3_1\",\"DSC3_2\"],\n",
    "                        method=\"wilcoxon\",reference=\"DSC3_1\",n_genes=30)\n",
    "sc.pl.embedding(dsc,color=['TIMP1','SERPINF1','HSPB1','DCN','HLA-A','HLA-B','HLA-C','B2M','CD81','C1S','C1R','C3',\n",
    "            'PPIA','UBB','UBC','APOE'],basis=\"X_umap\",cmap='viridis_r',vmin=0.5,frameon=False,palette='Set1')\n",
    "sc.pl.embedding(dsc,color=\"subclusters\",basis=\"X_umap\",groups=['DSC3_1','DSC3_2'],\n",
    "                palette='Set1')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17ea7975-00b9-4b33-8176-80e3b514ea31",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 04 Invasive EVTs counting on STOMICS\n",
    "\n",
    "import squidpy as sq\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "sq.gr.spatial_neighbors(adata, set_diag=True,coord_type=None,spatial_key='shift_spatial', \n",
    "                        delaunay=False,n_neighs=5)\n",
    "\n",
    "dsc=adata.obs.loc[adata.obs.celltype=='DSC',:]\n",
    "DSC4=dsc.loc[dsc.subtype.isin(['DSC4'])].index\n",
    "DSC3=dsc.loc[dsc.subtype.isin(['DSC3'])].index\n",
    "otherDSC=dsc.loc[~dsc.subtype.isin(['DSC3','DSC4'])].index\n",
    "\n",
    "## Searching spatial neighbors of DSCs\n",
    "def neighbor_search(adata, cells, spatial_key='spatial_connectivities'):\n",
    "    mask = adata.obs.index.isin(cells)\n",
    "    index_numbers = np.where(mask)[0]\n",
    "    dist_matrix = adata.obsp[spatial_key]\n",
    "    neighbors = [np.asarray([j for j in dist_matrix[i].indices if j != i]) for i in index_numbers]\n",
    "    neighbor_index, neighbor_subclass=[],[]\n",
    "    for i in neighbors:\n",
    "        neighbor_index.append([adata.obs.index[ids] for ids in i])\n",
    "\n",
    "    dsc_neighbors=[]\n",
    "    for i in neighbor_index:\n",
    "        dsc_neighbors+= i\n",
    "    return list(set(dsc_neighbors))\n",
    "\n",
    "## locate neighboring  cell  ids\n",
    "evt_ids = adata.obs.loc[adata.obs['subclass']=='EVT'].index\n",
    "dsc3_neighbors = neighbor_search(adata,dsc3)\n",
    "dsc3_evts = np.intersect1d(dsc3_neighbors,evt_ids)\n",
    "\n",
    "dsc4_neighbors = neighbor_search(adata,dsc4)\n",
    "dsc4_evts = np.intersect1d(dsc4_neighbors,evt_ids)\n",
    "\n",
    "otherdsc_neighbors = neighbor_search(adata,otherDSC)\n",
    "otherdsc_evts = np.intersect1d(otherdsc_neighbors,evt_ids)\n",
    "\n",
    "groups1 = list(set(dsc3_evts) - set(dsc4_evts) - set(otherdsc_evts))\n",
    "groups2 = list(set(dsc4_evts) - set(dsc3_evts) - set(otherdsc_evts))\n",
    "groups3 = list(set(otherdsc_evts) - set(dsc3_evts) - set(dsc4_evts))\n",
    "\n",
    "## iScore comparions\n",
    "evt=adata[adata.obs.subclass=='EVT',:].obs\n",
    "evt['group']='allEVTs'\n",
    "evt.loc[group1,'group']='nearDSC3'\n",
    "evt.loc[group2,'group']='nearDSC4'\n",
    "evt.loc[group3,'group']='otherDSC'\n",
    "sns.violinplot(data=evt, color='group',x='group',y='iScore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4395bafc-c80f-4b16-9c49-aec8e1bcc855",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 05 scRNA-seq Preprocessing in vitro decidualization\n",
    "\n",
    "import anndata as ad\n",
    "import os\n",
    "import scanpy as sc\n",
    "datadir=\"/path/to/cellranger/mAEA/\"\n",
    "files=os.listdir(datadir)\n",
    "adata_list=[]\n",
    "\n",
    "## merging adata\n",
    "for i in files:\n",
    "    tmp=sc.read_10x_mtx(os.path.join(datadir,i,\"/outs/filtered_feature_bc_matrix/\"))\n",
    "    tmp.obs['dataset']=i\n",
    "    tmp.var_names_make_unique(\"-\")\n",
    "    tmp.obs.index=i+\"_\"+pd.Series(tmp.obs.index)\n",
    "    adata_list.append(tmp)\n",
    "adata=ad.concat(adata_list,label=\"ID\")\n",
    "adata.obs['ID']=adata.obs['ID'].astype('str')\n",
    "\n",
    "## QC and integrating\n",
    "sc.pl.highest_expr_genes(adata, n_top=20, )\n",
    "sc.pp.filter_cells(adata, min_genes=200)\n",
    "sc.pp.filter_genes(adata, min_cells=0)\n",
    "adata.var['mt'] = adata.var_names.str.startswith('MT-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "sc.pl.violin(adata, ['n_genes_by_counts', 'total_counts', 'pct_counts_mt'],jitter=0.4, multi_panel=True)\n",
    "sc.pl.scatter(adata, x='total_counts', y='pct_counts_mt')\n",
    "sc.pl.scatter(adata, x='total_counts', y='n_genes_by_counts')\n",
    "adata = adata[adata.obs.n_genes_by_counts < 5000, :]\n",
    "adata = adata[adata.obs.pct_counts_mt < 20, :]\n",
    "\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5,)\n",
    "sc.pl.highly_variable_genes(adata,)\n",
    "adata.raw = adata\n",
    "adata = adata[:, adata.var.highly_variable]\n",
    "sc.pp.scale(adata, max_value=10)\n",
    "sc.tl.pca(adata, svd_solver='arpack')\n",
    "sc.pl.pca_variance_ratio(adata, log=True)\n",
    "sc.pp.neighbors(adata, n_neighbors=30, n_pcs=50)\n",
    "sc.tl.umap(adata,random_state=44)\n",
    "sc.tl.louvain(adata)\n",
    "sc.pl.umap(adata, color=['louvain','CNR1', 'ACTA2', 'IGFBP1','dataset'], use_raw=True,vmax='p99')\n",
    "adata.raw.to_adata().write(\"mAEA_processed.h5ad\")\n",
    "\n",
    "## Cell type identification \n",
    "\n",
    "dsc_markers=pd.read_table('/path/to/dsc_markers.tsv',sep=\"\\t\",index_col=0,header=0)\n",
    "set_a=dsc_markers.sort_values('DSC3_scores',ascending=False).index[:50]\n",
    "set_b=dsc_markers.sort_values('DSC4_scores',ascending=False).index[:50]\n",
    "set_a=[i for i in set_a if 'RPS' not in i and 'RPL' not in i and 'MT-' not in i]\n",
    "set_b=[i for i in set_a if 'RPS' not in i and 'RPL' not in i and 'MT-' not in i]\n",
    "sc.tl.score_genes(adata,gene_list=set_a,score_name=\"DSC3_score\",use_raw=True)\n",
    "sc.tl.score_genes(adata,gene_list=set_b,score_name=\"DSC4_score\",use_raw=True)\n",
    "sc.pl.umap(adata, color=['louvain','DSC4_score'], use_raw=False,vmin=-0.4,vmax=0.4)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88c27e85-089f-49bc-af86-307638bd8791",
   "metadata": {},
   "outputs": [],
   "source": [
    "##06 DEG mAEA vs EtOH\n",
    "import pandas as pd\n",
    "writer=pd.ExcelWriter(\"/path/to/mAEA_vs_EtOH.xlsx\")\n",
    "for i in adata.obs.clusters.unique():\n",
    "    adata2=adata[adata.obs.clusters==i,:]\n",
    "    sc.tl.rank_genes_groups(adata2,groupby='Condition',groups=['mAEA','EtOH'],reference='EtOH',method='wilcoxon')\n",
    "    df=sc.get.rank_genes_groups_df(adata2,group='mAEA').to_excel(writer,\n",
    "    sheet_name=\"{group}\".format(group='mAEA_vs_EtOH_'+str(i)),header=True,index=True)\n",
    "writer.save()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59df4611-c16e-4d82-a91f-fc7fcb0d3c9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 07 GSEA mAEA vs EtOH\n",
    "\n",
    "import pandas as pd\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "wdir='/path/to/gsea_ouptut/'\n",
    "gene_set_names = gp.get_library_name(organism='Human')\n",
    "print(gene_set_names)\n",
    "\n",
    "gene_set='GO_Biological_Process_2025'\n",
    "by='logfoldchanges'\n",
    "df1=pd.read_table(\"mAEA_vs_EtOH_DSC3.tsv\",sep='\\t',index_col=0,header=0)\n",
    "df2=pd.read_table(\"mAEA_vs_EtOH_DSC3.tsv\",sep='\\t',index_col=0,header=0)\n",
    "\n",
    "gene_rank=df1.loc[:,['gene',by]]\n",
    "gene_rank.sort_values(by=[by], inplace=True, ascending=False)\n",
    "res=gp.prerank(rnk=gene_rank, gene_sets=gene_set,permutation_num=1000,outdir=wdir+'/dsc3/')\n",
    "nes=res.res2d.sort_index(ascending=True)\n",
    "\n",
    "gene_rank=df2.loc[:,['gene',by]]\n",
    "gene_rank.sort_values(by=[by], inplace=True, ascending=False)\n",
    "res=gp.prerank(rnk=gene_rank, gene_sets=gene_set,permutation_num=1000,outdir=wdir+'/dsc4/')\n",
    "nes=res.res2d.sort_index(ascending=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
