{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "68f71c34",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import scipy\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import anndata as ad\n",
    "import seaborn as sns\n",
    "pd.set_option('display.max_columns', 100)\n",
    "pd.set_option('display.max_rows', 200)\n",
    "np.set_printoptions(suppress=True)\n",
    "#scv.settings.verbosity = 3\n",
    "#scv.settings.presenter_view = True\n",
    "#scv.set_figure_params('scanpy')\n",
    "pd.set_option('display.max_columns', 100)\n",
    "pd.set_option('display.max_rows', 200)\n",
    "np.set_printoptions(suppress=True)\n",
    "sc.set_figure_params(dpi=100, color_map = 'viridis_r')\n",
    "sc.settings.verbosity = 1\n",
    "plt.rcParams['pdf.fonttype']=42"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "675b3d6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## 01. QC & Doublets Removal using Scrublet with Default Settings \n",
    "wdir=\"/path/to/cellranger/\"\n",
    "files=os.listdir(wdir) ## list all sample ids\n",
    "dub_list,score_list=[],[]\n",
    "for i in files:\n",
    "    filtered= os.path.join(wdir,i,\"/outs/filtered_feature_bc_matrix.h5\")\n",
    "    c_filtered=sc.read_10x_h5(filtered, gex_only=True, backup_url=None)\n",
    "    c_filtered.obs['dataset']=i\n",
    "    c_filtered.obs['CB']=c_filtered_obs.index\n",
    "    c_filtered.obs.index=c_filtered.obs['dataset'].str.cat(c_filtered.obs['CB'],sep='_')\n",
    "    exp=c_filtered.n_obs/1000*0.4/100 ## setting the priors for 10K cells per samples\n",
    "    print('Sample {} Number of cells in cell list: {}'.format(i, c_filtered.n_obs))\n",
    "    print('Sample {} Number of genes in gene list: {}'.format(i, c_filtered.n_vars))\n",
    "    sc.external.pp.scrublet(c_filtered,expected_doublet_rate=exp,log_transform=True,n_prin_comps=30)\n",
    "    score_list.append(c_filtered.obs['doublet_score'])\n",
    "    dub_list.append(c_filtered.obs['predicted_doublet'])\n",
    "\n",
    "score_list=pd.concat(score_list)\n",
    "dub_list=pd.concat(dub_list)\n",
    "dubs=dub_list[dub_list==1].index\n",
    "\n",
    "adata=sc.read(\"/path/to/h5ad/integrated_gex.h5ad\")\n",
    "\n",
    "adata = adata[(adata.obs.nCount_RNA<1000)&(adata.obs.nCount_RNA<50000)&\n",
    "              (adata.obs.nCount_ATAC>1000)&(adata.obs.nCount_ATAC<100000)&\n",
    "              (adata.obs.nFeature_RNA>400)&(adata.obs.nCount_ATAC<100000)&\n",
    "              (adata.obs[\"percent.mt\"]<0.2)&(adata.obs['TSS.enrichment'].gt(1.0))& \n",
    "              (adata.obs['nucleosome_signal'].lt(2.0)),:]                   ## QC\n",
    "\n",
    "adata_filtered = adata[[i for i in adata.obs.index if i not in dubs],:] ## removal all predicted doublets \n",
    "adata_filtered.write(\"/path/to/h5ad/integrated_gex_filtered.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a83c607",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## 02. UMAP plotting\n",
    "adata=sc.read(\"/tank/data2/cw/snPlacenta/download/scPlacenta_host.h5ad\")\n",
    "colors = {\n",
    "  \"SCT\" : \"#3F84AA\",\n",
    "  \"Epi\" : \"#C9EBFB\",\n",
    "  \"DSC\" : \"#65A83E\",\n",
    "  \"VCT\" : \"#8870ad\",\n",
    "  \"dNK\" :  \"#C3C388\",#\n",
    "  \"M\" : \"#c9a997\",\n",
    "  \"EVT\" : \"#F397C0\",#\n",
    "  \"FB\" :  \"#C594BF\",#                 \n",
    "  \"HB\" :  \"#FBBE92\",#\n",
    "  \"PV\" :  \"#9e6762\",#                              \n",
    "  \"T\" :  \"#DFCDE4\",#\n",
    "  \"cDC\" :  \"#989898\",#                               \n",
    "  \"Ery\" :  \"#c9a997\",#   \n",
    "  \"B\" : \"#B51D8D\",\n",
    "'LEC':\"#FACB12\",\n",
    "'eS':\"#8EC792\",\n",
    "'fVEC':\"#eda450\",\n",
    "'mVEC':\"#EF4E22\",\n",
    "\"EC\" : \"#EF4E22\",\n",
    "'cili':\"#139992\"}\n",
    "\n",
    "\n",
    "colPalette = [colors[i] for i in sorted(np.unique(adata.obs['minor_class']))]\n",
    "\n",
    "\n",
    "colPalette = [colors[i] for i in sorted(np.unique(adata.obs['minor_class']))]\n",
    "sc.pl.scatter(adata,color='minor_class',basis=\"umap\",legend_loc='on data',\n",
    "              palette=colPalette,size=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56b0a0d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 03. Cell type composition across gestational ages\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import re\n",
    "from matplotlib.colors import ListedColormap\n",
    "\n",
    "## Piechart\n",
    "def func(pct, allvals):\n",
    "    absolute = int(np.round(pct/100.*np.sum(allvals)))\n",
    "    return \"{:.1f}%\".format(pct)\n",
    "\n",
    "#adata=sc.read(\"/tank/data2/cw/snPlacenta/download/scPlacenta_host.h5ad\")\n",
    "stages=adata.obs.gestational_age_group.unique()\n",
    "pal=sns.color_palette('tab20', 15)\n",
    "fig, axs = plt.subplots(ncols=len(stages), nrows=1, figsize=(25,5))\n",
    "pal=pal.as_hex()\n",
    "order=adata.obs.major_class.value_counts().index.astype('str')\n",
    "\n",
    "for col,j in enumerate(stages):\n",
    "    fig.add_subplot(axs[col])\n",
    "    df=adata.obs.loc[adata.obs['gestational_age_group']==j,'major_class'].value_counts().loc[order]\n",
    "    wedges=axs[col].pie(df,labels=df.index,colors=pal)\n",
    "    plt.axis('off')\n",
    "\n",
    "\n",
    "## Stacked Area Plot\n",
    "\n",
    "gws=pd.DataFrame(index=adata.obs.gestational_week.unique().astype('str'))\n",
    "gws['gw'] = [int(re.findall(r'\\d+', x)[0]) for x in gws.index]\n",
    "gws['ga_subgroup']=pd.cut(gws['gw'], bins=[5, 9, 13, 20, 25, 40],labels=[\"early1st\", \"late1st\", \"early2nd\",'late2nd','term'])\n",
    "adata.obs['ga_subgroup']=adata.obs['gestational_week'].map(dict(gws['ga_subgroup']))\n",
    "df=adata.obs[['ga_subgroup','major_class']].value_counts().reset_index()\n",
    "df['proportion'] = df.groupby('ga_subgroup')[0].transform(lambda x: x / x.sum())\n",
    "\n",
    "\n",
    "\n",
    "colPalette = [colors[i] for i in sorted(np.unique(adata.obs['major_class']))]\n",
    "cmap = ListedColormap(colPalette)\n",
    "\n",
    "\n",
    "\n",
    "pivot_df = df.pivot(index='ga_subgroup', columns='major_class', values='proportion')\n",
    "pivot_df = pivot_df.fillna(0)\n",
    "order=['early1st','late1st','early2nd','late2nd','term']\n",
    "pivot_df=pivot_df.loc[order]\n",
    "pivot_df.plot.area(\n",
    "    colormap=cmap,\n",
    "    figsize=(10,6),\n",
    "    linewidth=0)\n",
    "\n",
    "plt.xlabel(\"Time point\")\n",
    "plt.ylabel(\"Proportion\")\n",
    "plt.title(\"Cluster Composition Over Time\")\n",
    "plt.legend(title='Cluster', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.grid(False)\n",
    "plt.tight_layout()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25e4b5ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 04 Gene expression dotplot for markers\n",
    "target_gene=['TP63','TENM3','HLA-G','PRG2','ERVW-1','HOPX','DKK1','LUM','PGR','IGF1',\"BMP5\",'PDPN',\n",
    "             \"ACTA2\",\"AFF2\",\"POU5F1\",\"PAEP\",\"FOXJ1\",\"TP73\",\"PECAM1\",'VWF','MEOX2','PROX1','CCL21','IGHM',\n",
    "             'BCL11A',\"CD3D\",\"CD3G\",\"GNLY\",\"NKG7\",\"MS4A7\",\"CD14\",'MRC1','LYVE1','HLA-DRA','CD74',\"HBA1\",\"HBB\"]\n",
    "categories_order=['VCT',\"EVT\",\"SCT\",\"DSC\",'eS',\"FB\",\"PV\",\"Epi\",'cili',\"mVEC\",'fVEC','LEC',\n",
    "                  \"B\",\"T\",\"dNK\",\"M\",\"HB\",\"cDC\",\"Ery\"]\n",
    "sc.pl.dotplot(adata,groupby='minor_class',var_names=target_gene,standard_scale='var',cmap='coolwarm',\n",
    "              categories_order=categories_order,dot_max=0.6,use_raw=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6371d36",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 05 piechart of maternal vs fetal origin per cell type \n",
    "\n",
    "def func(pct, allvals):\n",
    "    absolute = int(np.round(pct/100.*np.sum(allvals)))\n",
    "    return \"{:.1f}%\".format(pct)\n",
    "x=adata.obs.loc[adata.obs.major_class==\"FB\",\"origin\"].value_counts()\n",
    "plt.pie(x,labels=x.index,autopct=lambda pct: func(pct, x),colors=['dodgerblue','deeppink','grey'])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01527f21",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 06 Heatmap plotting on motif activity by chromVar\n",
    "chromvar=sc.read(\"/path/to/chromVar.h5ad\")\n",
    "chromvar=chromvar[adata.obs.index,:]\n",
    "chromvar.obs=adata.obs\n",
    "scv.set_figure_params('scanpy')\n",
    "sc.tl.rank_genes_groups(chromvar, groupby='minor_class',method='wilcoxon')\n",
    "\n",
    "motif_ids=pd.read_csv(\"/path/to/chromvar.motif.annotation.txt\",sep=\"\\t\",header=0,index_col=2)\n",
    "chromvar.var=motif_ids\n",
    "chromvar.var['MA']=chromvar.var.index\n",
    "chromvar.var.index=chromvar.var.jaspar2020\n",
    "categories_order=['VCT',\"EVT\",\"SCT\",\"STM\",\"FB\",\"PV\",\"Epi\",\"Endo\",\"B\",\"T\",\"dNK\",\"M\",\"HB\",\"cDC\",\"Erythrocyte\"]\n",
    "TFs=['TP63','TEAD3',\"TEAD1\",'GATA3',\"GCM1\",\"TFAP2B\",\"TFAP2A\",\n",
    "     \"HAND2\",\"HOXA10\",\"EBF1\",\"HNF1B\",\"HOXB4\",\"ETV1\",\"IRF4\",\"IKZF1\",\"RUNX2\",\"ETS1\",'RUNX2',\"SPI1\",\"ELF1\",\"ETV6\",\n",
    "     \"GATA1::TAL1\"]\n",
    "sc.pl.matrixplot(chromvar, TFs, categories_order=categories_order,groupby='subclass',dendrogram=False,\n",
    "              cmap=\"bwr\",standard_scale='var')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cac98e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 07 Heatmap of Enhancers with Differential Accessibility\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "plt.rcParams['pdf.fonttype']=42\n",
    "\n",
    "enhancer=sc.read(\"/path/to/F5.enhancer.h5ad\")\n",
    "enhancer_anno=pd.read_table(\"/path/to/F5.enhancer.anno\",sep=\"\\t\",header=0,index_col=None)\n",
    "enhancer_anno=pd.concat((enhancer.var.reset_index(),enhancer_anno),1)\n",
    "enhancer_anno.index=enhancer_anno.iloc[:,0]\n",
    "enhancer_anno=enhancer_anno.drop(\"index\",1)\n",
    "\n",
    "sc.tl.rank_genes_groups(enhancer, groupby='minor_class',method='wilcoxon')\n",
    "\n",
    "\n",
    "order = ['VCT',\"EVT\",\"SCT\",'DSC','eS',\"FB\",'PV',\"Epi\",\"cili\",\n",
    "        \"mVEC\",\"fVEC\",\"LEC\",\"B\",\"T\",\"dNK\",\"M\",'HB','cDC',\"Ery\"]\n",
    "\n",
    "rgg = enhancer.uns['rank_genes_groups']\n",
    "\n",
    "for key in ['names', 'scores', 'logfoldchanges', 'pvals', 'pvals_adj']:\n",
    "    arr = rgg[key]\n",
    "    reordered = np.array([arr[name] for name in desired_order])\n",
    "    reordered = np.core.records.fromarrays(reordered, names=order)\n",
    "    enhancer.uns['rank_genes_groups'][key] = reordered               # Reorder columns based on desired order\n",
    "\n",
    "\n",
    "sc.pl.rank_genes_groups_matrixplot(enhancer, n_genes=10, groupby=\"minor_class\", gene_symbols='name',\n",
    "                                key='rank_genes_groups',use_raw=False,dendrogram=False,standard_scale=\"var\",\n",
    "                                categories_order=order,swap_axes=True,save=\"enhancer_heatmap.pdf\")\n",
    "\n",
    "names_array = enhancer.uns['rank_genes_groups']['names']  \n",
    "group_names = names_array.dtype.names \n",
    "n_genes = names_array.shape[0]\n",
    "\n",
    "## map enhancer to annotation\n",
    "gene_symbol = []\n",
    "\n",
    "for group in group_names:\n",
    "    varnames = names_array[group]  # top genes (as var_names) # Map to gene symbols using adata.var\n",
    "    symbols = [enhancer.var['gene_name'].get(gene, gene) for gene in varnames]\n",
    "    gene_symbol.append(symbols)\n",
    "\n",
    "enhancer_df=sc.get.rank_genes_groups_df(enhancer,group=None,key=\"wilcoxon\")\n",
    "enhancer_df=enhancer_df.merge(enhancer_anno,left_on=\"names\",right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faa098bc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sc",
   "language": "python",
   "name": "sc"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
